crates/bevy_reflect/src/impls/smol_str.rs

```diff
@@ -0,0 +1,134 @@
+use std::{
+    any::Any,
+    hash::{Hash, Hasher},
+};
+
+use smol_str::SmolStr;
+
+use crate::{
+    utility::{reflect_hasher, NonGenericTypeInfoCell},
+    FromReflect, FromType, GetTypeRegistration, Reflect, ReflectDeserialize, ReflectFromPtr,
+    ReflectMut, ReflectOwned, ReflectRef, ReflectSerialize, TypeInfo, TypeRegistration, Typed,
+    ValueInfo,
+};
+
+impl Reflect for SmolStr {
+    fn type_name(&self) -> &str {
+        std::any::type_name::<Self>()
+    }
+
+    fn get_represented_type_info(&self) -> Option<&'static TypeInfo> {
+        Some(<Self as Typed>::type_info())
+    }
+
+    fn into_any(self: Box<Self>) -> Box<dyn Any> {
+        self
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn into_reflect(self: Box<Self>) -> Box<dyn Reflect> {
+        self
+    }
+
+    fn as_reflect(&self) -> &dyn Reflect {
+        self
+    }
+
+    fn as_reflect_mut(&mut self) -> &mut dyn Reflect {
+        self
+    }
+
+    fn apply(&mut self, value: &dyn Reflect) {
+        let value = value.as_any();
+        if let Some(value) = value.downcast_ref::<String>() {
+            *self = SmolStr::new(value);
+        } else {
+            panic!("Value is not a {}.", std::any::type_name::<Self>());
+        }
```

Comment:
I think the idea of allowing `apply` to be called with a `String` is a good idea (and maybe something we should do for `Cow<'static, str>`). However, we should probably also make it easy to apply from another `SmolStr` (i.e. `Self`).

Reply:
I wanted to have the underlying type a string, from my understanding, apply is used exclusively with:

```rs
// This would create an underlying String
let a = SmolStr::new("toto").as_reflect();
let other = SmolStr::new("tata");
// So that needs to check on a String
other.apply(&a);
```

In this current implementation strategy, I'm failing to see when a dyn Reflect containing a SmolStr could be constructed in the first place.

But that probably betray a lack of understanding: I wanted to rely on String to be able to inspect its content ; I think currently it's not quite correct because I should:
- actually store a String (https://github.com/bevyengine/bevy/pull/8771/files/cf33bdcb4c638789bcfd19012dca047fe9ad16c7#r1221078886)
- I'm not sure how to forward the string reflect capabilities to my &dyn Reflect (representing a SmolStr).

Reply:
> I added a first check on SmolStr, then a fallback to String 🤷 ; still not sure how that could happen though.

It could occur like this:

```rust
let mut value = SmolStr::new("Hello");

let patch = String::from("Goodbye");
value.apply(&patch);

assert_eq!("Goodbye", value);
```

This is just a convenience for users to interact with a reflected `SmolStr`. If you don't want or need this behavior, then it is not necessary by any means.